govuk_chat_evaluation/rag_answers/custom_deepeval/metrics/context_relevancy/context_relevancy.py

```diff
@@ -0,0 +1,202 @@
+                self,
+                steps=[
+                    f"Truths: {prettify_list(truth_collection.truths)}",
+                    f"Information Needs:\n{prettify_list(information_needs_collection.information_needs)}",
+                    f"Verdicts:\n{prettify_list(verdict_collection.verdicts)}",
+                    f"Score: {self.score}\nReason: {self.reason}",
+                ],
+            )
+
+            return self.score
+
+    async def _generate_truths(self, retrieval_context: list[str]) -> TruthCollection:
+        prompt = self.evaluation_template.truths(
+            retrieval_context=retrieval_context,
+        )
+        return await self._generate_result_from_model(prompt, schema=TruthCollection)
+
+    async def _generate_information_needs(
+        self, input: str
+    ) -> InformationNeedsCollection:
+        prompt = self.evaluation_template.information_needs(input=input)
+        return await self._generate_result_from_model(
+            prompt, schema=InformationNeedsCollection
+        )
+
+    async def _generate_verdicts(
+        self,
+        information_needs_collection: InformationNeedsCollection,
+        truth_collection: TruthCollection,
+    ) -> VerdictCollection:
+        if len(information_needs_collection.information_needs) == 0:
+            return VerdictCollection(verdicts=[])
+
+        extracted_truths = truth_collection.extracted_truths()
+        prompt = self.evaluation_template.verdicts(
+            information_needs=information_needs_collection.information_needs,
+            extracted_truths=extracted_truths,
+        )
+
+        return await self._generate_result_from_model(prompt, schema=VerdictCollection)
+
+    def _calculate_score(self, verdicts: VerdictCollection) -> float:
+        score = verdicts.score_verdicts()
+
+        return 0 if self.strict_mode and score < self.threshold else score
+
+    async def _generate_reason(
+        self, input: str, verdict_collection: VerdictCollection
+    ) -> str | None:
+        if self.include_reason is False:
+            return None
+
+        unmet_needs = verdict_collection.unmet_needs()
+        score = float(round(self.score or 0.0, 2))
+
+        prompt = self.evaluation_template.reason(
+            unmet_needs=unmet_needs,
+            input=input,
+            score=score,
+        )
+
+        result = await self._generate_result_from_model(prompt, schema=ScoreReason)
+        return result.reason if isinstance(result.reason, str) else str(result.reason)
+
+    async def _generate_result_from_model(
+        self, prompt: str, schema: Type[SchemaType]
+    ) -> SchemaType:
+        if self.using_native_model:
+            result, cost = await self.model.a_generate(prompt, schema=schema)
+            self._increment_cost(cost)
+            if isinstance(result, schema):
+                return result
+            else:
+                raise TypeError("Generated result does not match the expected schema.")
+        else:
```

Comment:
You shouldn't need an else here since the if above will have exited the function

Reply:
getting rid of this depends on this https://github.com/alphagov/govuk-chat-evaluation/pull/127#discussion_r2490469981